File: projects/engineering.purdue.edu/scriptingwithobjects/swocode/chap4/MultiLineRegex.py
#!/usr/bin/env python

### MultiLineRegex.py

import sys                                                           #(A)
import re                                                            #(B)

if len(sys.argv) is not 2:                                           #(C)
    sys.exit( "Name the XML file in the command line")               #(D)

try:                                                                 #(E)
    FILE = open( sys.argv[1], 'r' )                                  #(F)
except IOError, e:                                                   #(G)
    print "Cannot open file: ", e                                    #(H)
    sys.exit(1)                                                      #(I)

the_whole_file = FILE.read()                                         #(J)

regex = r'<listing>(.*?)<\/listing>'                                 #(K)
matches = re.findall( regex, the_whole_file, re.DOTALL )             #(L)

regex =  re.compile( r'''<first>(.*)<\/first>.*?    # get first name #(M)
                         <last>(.*)<\/last>.*?      # get last name  #(N)
                         <phone.*>(.*)<\/phone>''', # get phone num  #(O)
                     re.DOTALL | re.VERBOSE )                        #(P)
phonelist = {}                                                       #(Q)
for listing in matches:                                              #(R)
    m = re.search( regex, listing )                                  #(S)
    if m is not None:                                                #(T)
        phonelist["%s, %s" % (m.group(2), m.group(1))] = m.group(3)  #(U)

for item in sorted( phonelist.items() ):                             #(V)
    print "      %-20s\t%15s" % item                                 #(W)

File: projects/engineering.purdue.edu/scriptingwithobjects/swocode/chap4/MultiLineRegex.pl
#!/usr/bin/perl -w

### MultiLineRegex.pl

use strict;                                                          #(A)
use English;                                                         #(B)

die "needs one command line arg, the name of the XML file" 
    unless @ARGV == 1;                                               #(C)

$INPUT_RECORD_SEPARATOR = undef;                                     #(D)

my $the_whole_file = <>;                                             #(E)

my @listings = $the_whole_file =~ /<listing>(.*?)<\/listing>/sg;     #(F)

my %phonelist;                                                       #(G)

map {m/<first>(.*)<\/first>.*?               # get first name        #(H)
       <last>(.*)<\/last>.*?                 # get last name         #(I)
       <phone.*>(.*)<\/phone>                # get phone num         #(J)
      /sx; $phonelist{"$2, $1"} = $3 } @listings;                    #(K)

my $template = "@<<<<<<<<<<<<<<<<<<     @>>>>>>>>>>>>>>";            #(L)
foreach (sort keys %phonelist) {                                     #(M)
    formline( $template, $_, $phonelist{$_} );                       #(N)
    print $ACCUMULATOR, "\n";                                        #(O)
    $ACCUMULATOR = "";                                               #(P)
}
